experimental/structs.md

```diff
@@ -0,0 +1,57 @@
+# [JavaScript Structs: Fixed Layout Objects and Some Synchronization Primitives](proposal-structs)
+
+# Structs
+
+```js
+interface Struct <: Node {
+    shared: boolean;
+    id: Identifier | null;
+    superClass: Expression | null;
+    body: StructBody;
+}
+```
+
+## StructDeclaration
+```js
+interface StructDeclaration <: Struct, Declaration {
+    type: "StructDeclaration";
+    id: Identifier;
+}
+```
+
+## StructBody
+
+```js
+interface StructBody <: Node {
+    type: "StructBody";
+    body: [ MethodDefinition | PropertyDefinition | StaticBlock ]
+}
+```
+
+# Statements
+
+## UnsafeBlockStatement
+
+```js
+interface UnsafeBlockStatement <: BlockStatement {
+    type: "UnsafeBlockStatement";
```

Comment:
An alternative design is to add `shared: boolean` to the `BlockStatement` node. However, in ESTree the `BlockStatement` node is also used in FunctionDeclaration, I think it is not worthy to add a boolean property to such a popular AST node when we can confirm it is always `false` in this situation.

Reply:
> An alternative design is to add `shared: boolean`

I assume you meant `unsafe: boolean` since "shared" applies to structs not blocks?

I agree a separate type makes sense, although it's slightly unfortunate that traverse algorithms will have to add a new node type even though it should be treated 1:1 as any other blocks.

Reply:
Yes, I was meant for `unsafe: boolean`.
> even though it should be treated 1:1 as any other blocks.
Right, we already have `StaticBlock` as a precedent case, which also extends from `BlockStatement`.

An alternative design is to model `UnsafeBlockStatement` as a wrapper of `BlockStatement`:
```js
interface UnsafeBlockStatement <: Statement {
  body: BlockStatement
}
```
The Pros of this design is that linters will probably don't have to specifically handle `UnsafeBlockStatement`. 
The cons is the extra hierarchy.

Reply:
I think a new node is better because it means we  can easily limit any future nodes that might only be allowed in unsafe blocks.